core/src/main/java/org/infinispan/CachePublisher.java

```diff
@@ -0,0 +1,222 @@
+    */
+   <R> SegmentPublisherSupplier<R> keyPublisher(SerializableFunction<? super Publisher<K>, ? extends Publisher<R>> transformer);
+
+   /**
+    * Performs the given <b>transformer</b> on data in the cache, resulting in multiple values. If a single
+    * value is desired, the user should use {@link #entryReduction(Function, Function)}
+    * instead as it can optimize some things. Depending on the <b>deliveryGuarantee</b> the <b>transformer</b> may be
+    * invoked <b>1..numSegments</b> times per node. Results from a given node will retrieve values up to
+    * {@code batchSize} values until some are consumed.
+    * <p>
+    * For example when using RxJava and using an intermediate operation such as
+    * {@link io.reactivex.rxjava3.core.Flowable#switchIfEmpty(Publisher)} this can add elements if the given Publisher
+    * is empty, and it is very possible that a segment may not have entries and therefore may add the elements the
+    * switched Publisher returns multiple times.
+    * <p>
+    * Methods that add elements to the returned Publisher are fine as long as they are tied to a specific entry, for
+    * example {@link io.reactivex.rxjava3.core.Flowable#flatMap(io.reactivex.rxjava3.functions.Function)} which can reproduce
+    * the same elements when provided the same input entry from the cache.
+    *
+    * @param transformer transform the given stream of data into something else (requires non null)
+    * @param <R> return value type
+    * @return Publisher that when subscribed to will return the results and notify of segment completion if necessary
+    */
+   <R> SegmentPublisherSupplier<R> entryPublisher(Function<? super Publisher<CacheEntry<K, V>>, ? extends Publisher<R>> transformer);
+
+   /**
+    * Same as {@link #entryPublisher(Function)} except that the
+    * Function must also implement <code>Serializable</code>
+    * <p>
+    * The compiler will pick this overload for lambda parameters, making them <code>Serializable</code>
+    * @param transformer
+    * @return
+    * @param <R>
+    */
+   <R> SegmentPublisherSupplier<R> entryPublisher(SerializableFunction<? super Publisher<CacheEntry<K, V>>, ? extends Publisher<R>> transformer);
+
+   /**
+    * Similar to {@link CacheStream#forEach(SerializableBiConsumer)} in that it allows for the <b>transformer</b> and
+    * the <b>consumer</b> to be both ran on remote nodes allowing for lowest latency updates. Note that the operation
+    * will usually be ran on the primary owner, but there is no guarantee of this, especially during a rehash. If a rehash
+    * occurs the consumer may be ran more than once, so it is highly recommended to ensure these operations are idempotent
+    * if possible.
+    * @param transformer
+    * @param consumer
+    * @return
+    * @param <R>
+    */
+   <R> CompletionStage<Void> performRemotely(Function<? super Publisher<CacheEntry<K, V>>, ? extends Publisher<R>> transformer,
```

Comment:
This method is the only method that requires a bit more of implementation specifics around it to allow the injection of the cache, which the `ClusterPublisherManager` doesn't currently support. But I think it would be useful as users today may use the `Stream.forEach` method.